FR #1090: New table demo - Option 1

rileyjenk opened the pull request.

rileyjenk committed:
Structural Changes to current paths to allow for additional tables.

rileyjenk committed:
Additional Table Detail...very examplish only.

rileyjenk committed:
More documentation on purpose of table and a top level doc describing data model.

flanakin commented on specification/tables/cost_and_usage/columns/availabilityzone.md:
I would keep the columns where they are today. New datasets should pull from the global set of columns rather than duplicate column definitions. This will be important when we reuse concepts.

mike-finopsorg replied:
@flanakin I agree with the reuse, I just wanted to build out your approach

* All columns for all datasets live in the same folder. 
* We use MDPP includes to select from columns to include in the relevant sections of the document. Allowing the document to have the primary FOCUS dataset definition, followed by adjacent datasets selecting the relevant columns (including some of those in the primary FOCUS dataset)
* Only one set of attributes to be included in the document covering attributes used in all datasets


does this match what you are saying?

mike-finopsorg replied:
If that's right I have a couple of questions for the group

1. Are we ok for the structure of the document to not be reflective of the structure of the filesystem/files. 
2. Inside the document (not files in git): Do we duplicate the definition of a column in the FOCUS dataset when it's used in an adjacent dataset or do we link/reference it back to the definition used in the FOCUS dataset?

flanakin replied:
I was thinking there'd be a single data dictionary with columns and attributes, then separate datasets that have a single table listing applicable columns and attributes.

Hopefully, we won't need to duplicate content in the spec. We just need to confirm whether cross-column requirements make sense in this model or if they'd need to change.

mike-finopsorg replied:
I guess this is a step away from what we do today which the list of columns is the FOCUS data spec. We would instead move to have the list of columns cover all data columns generically and then a section of the specification to state which columns go to which dataset (FOCUS or otherwise). 

I think this works ok, but would require us to all agree and understand this approach.

flanakin commented on specification/tables/cost_and_usage/purpose.md:
```diff
@@ -0,0 +1,2 @@
+The Cost and Usage table is a record of the costs and usage of resources or services in a given billing period. It is used to track and analyze the financial aspects of resource consumption.
```
Can we refer to these as "datasets" rather than "tables"?
```suggestion
The Cost and Usage dataset is a record of the costs and usage of resources or services in a given billing period. It is used to track and analyze the financial aspects of resource consumption.
```

flanakin commented on specification/tables/invoices/columns/invoiceid.md:
This is the perfect example of a duplicated column. If we maintain columns at a spec level, then we can reuse the definition. And they should ideally maintain the definition across datasets or it would be confusing.

That said, we will need to look at how cross-column requirements are applied. I'm thinking we want to avoid duplicating those, however.

flanakin commented on specification/tables/invoices/columns/invoiceid.md:
```diff
@@ -0,0 +1,42 @@
+
+The InvoiceId column adheres to the following requirements:
+
+* InvoiceId is RECOMMENDED to be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
```
I would expect to see a table of columns that should be in each dataset with their required level of inclusion. InvoiceId may be RECOMMENDED in cost dataset, but would be a MUST in the invoice dataset.

shawnalpay committed:
Merge branch 'working_draft' into newTableDemo

shawnalpay commented:
@rileyjenk What are the next steps on this PR?  The PDF builder is broken, which I would expect, given that this substantively changes the nature of the PDF generation.  What, if anything, do you need from the group to move this forward?  Or has it served its purpose and will not be merged?

shawnalpay commented:
We discussed this PR in Jul 28 Maintainers.  While we understand the utility of this PR, it would be helpful for there to be a summary of the two possible paths forward.  @mike-finopsorg took AI #1225 to bring that back to the Maintainers next week.

mike-finopsorg commented:
I have created https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1229 which demonstrates how a single column library with FOCUS dataset definitions would work as an alternative option to this PR for review of the maintainers.

mike-finopsorg committed:
Updated option 1 of supporting Adjacent datasets in the spec
Updated vendored markdown-pp package to support per top level heading TOC depth
Updated build system to support new folder layout
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented:
I have pushed a heap of changes to this PR to make it build correctly, including some patches to the vendored package (markdown-pp) and build scripts.

mike-finopsorg committed:
Merge branch 'working_draft' into newTableDemo

shawnalpay commented:
> I have pushed a heap of changes to this PR to make it build correctly, including some patches to the vendored package (markdown-pp) and build scripts.


Thanks for pulling this together @mike-finopsorg!  This looks really good.  For quick reference for everyone, **[here](https://github.com/user-attachments/files/21623189/spec.pdf)** is the PDF showing the output.  Notice that the columns are listed underneath the dataset names in the Table of Contents, but Mike has handled for suppressing that same depth for other ToC sections.

shawnalpay committed:
Merge branch 'working_draft' into newTableDemo

mike-finopsorg commented:
Closed with https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1337 superseding
